lib/TPP/ConvertLinalgToXsmm.cpp

```diff
@@ -260,6 +301,101 @@ struct ConvertGenericToUnaryRelu : public OpRewritePattern<linalg::GenericOp> {
+  case BroadCastType::COL:
+    return (operandIdx == 0) ? xsmm::BinaryFlags::BCAST_COL_IN_0
+                             : xsmm::BinaryFlags::BCAST_COL_IN_1;
+  default:
+    return xsmm::BinaryFlags::NONE;
+  }
+}
+
+// Convert linalg.generic to xsmm binary add op.
+struct ConvertGenericToBinaryAdd : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Value> operands;
+    if (!genericOp.hasBufferSemantics() ||
+        !structured_match::utils::isTwoDAddOp(genericOp, &operands) ||
+        operands.size() != 3) {
+      return failure();
+    }
+
+    auto lhs = operands[0];
+    auto rhs = operands[1];
+    auto output = operands[2];
+    Type outputType = output.getType();
+
+    auto stridesOnLhs = utils::getStaticStrides(lhs);
+    auto stridesOnRhs = utils::getStaticStrides(rhs);
+    auto stridesOnOutput = utils::getStaticStrides(output);
+
+    if (failed(stridesOnLhs) || failed(stridesOnRhs) || failed(stridesOnOutput))
+      return failure();
+    if (stridesOnLhs->back() != 1 || stridesOnRhs->back() != 1 ||
+        stridesOnOutput->back() != 1) {
+      return failure();
+    }
+
+    BinaryInfo binaryInfo;
+    binaryInfo.m = outputType.cast<ShapedType>().getShape()[0];
+    binaryInfo.n = outputType.cast<ShapedType>().getShape()[1];
+    binaryInfo.ldiLhs = stridesOnLhs->front();
+    binaryInfo.ldiRhs = stridesOnRhs->front();
+    binaryInfo.ldo = stridesOnOutput->front();
+
+    OpOperand *lhsOperand = getOperandFromValue(genericOp, lhs);
+    auto broadCastFlagLhs = getBroadCastBinaryFlagFromMap(
+        genericOp.getMatchingIndexingMap(lhsOperand), /*operandIdx=*/0);
+    if (failed(broadCastFlagLhs))
+      return failure();
+
+    OpOperand *rhsOperand = getOperandFromValue(genericOp, rhs);
+    auto broadCastFlagRhs = getBroadCastBinaryFlagFromMap(
+        genericOp.getMatchingIndexingMap(rhsOperand), /*operandIdx=*/1);
+    if (failed(broadCastFlagRhs))
+      return failure();
+
+    auto flagLhs =
+        xsmm::BinaryFlagsAttr::get(rewriter.getContext(), *broadCastFlagLhs);
+    auto flagRhs =
+        xsmm::BinaryFlagsAttr::get(rewriter.getContext(), *broadCastFlagRhs);
+
+    // Spaghetti code to handle 'NONE' as it conflicts with other flags; we
+    // cannot add it if at least the RHS or the LHS is not 'NONE'. Maybe the
+    // best solution is to get rid of it.
+    ArrayAttr flags;
+    if (flagLhs.getValue() != xsmm::BinaryFlags::NONE &&
+        flagRhs.getValue() != xsmm::BinaryFlags::NONE) {
+      flags = rewriter.getArrayAttr({flagLhs, flagRhs});
+    } else if (flagLhs.getValue() != xsmm::BinaryFlags::NONE) {
+      flags = rewriter.getArrayAttr({flagLhs});
+    } else if (flagRhs.getValue() != xsmm::BinaryFlags::NONE) {
+      flags = rewriter.getArrayAttr({flagRhs});
+    }
```

Comment:
Doesn't this need one more default case to add NONE?

Reply:
yes..